Course Dashboard: Reset Dates for PLS

mumer92 opened the pull request:
### Description

[LEARNER-7825](https://openedx.atlassian.net/browse/LEARNER-7825)

As a learner using edX's mobile app, I want to adjust my schedule when I miss an assignment date so I can manage my important course dates and stay on track

When an assignment date becomes past due, display the reset dates banner on both Course home and Dates page

- Banner text: It looks like you missed some important deadlines based on our suggested schedule.
- Button text: Shift due date

- Shift due dates success message: "Your dates have been successfully shifted."

- Use native platform UI and timing for the toast messages for Android and Apple

- Shift due dates error message: “Your dates could not be shifted. Please try again.”

- Use native platform UI and timing for the toast messages for Android and Apple

- Banner is not dismissible

### How to test this PR

You can return appropriate status from following lines to check the behaviours of Banner text and button

https://github.com/edx/edx-app-ios/blob/a2d67e6bf522bba7694597efea0fa1e7486054bb/Source/CourseDates.swift#L217-L229

### Screenshot
<img width="370" alt="Screenshot 2020-09-16 at 1 59 35 PM" src="https://user-images.githubusercontent.com/960241/93315801-e738ec80-f824-11ea-919a-95a5232bf7ab.png">

saeedbashir commented on Source/CourseOutlineViewController.swift:
```diff
@@ -156,8 +157,25 @@ public class CourseOutlineViewController :
+                break
+                
+            case .failure(let error):
+                print(error)
```
Instead of `print`, I think `Logger` is more useful here.

saeedbashir commented on Source/CourseOutlineViewController.swift:
```diff
@@ -156,8 +157,25 @@ public class CourseOutlineViewController :
-        stream.extendLifetimeUntilFirstResult (success : { [weak self] (rootID, block) in
+        let courseOutlineStream = joinStreams(courseQuerier.rootID, courseQuerier.blockWithID(id: blockID))
+        
+        let request = CourseDatesAPI.courseDeadlineInfoRequest(courseID: courseID)
+        let courseDeadlineStream = environment.networkManager.streamForRequest(request)
+        courseDeadlineLoader.backWithStream(courseDeadlineStream)
+        
+        courseDeadlineStream.listen(self) { [weak self] result in
+            switch result {
+            case .success(let courseDeadline):
+                self?.loadCourseDatesResetView(courseDeadline: courseDeadline)
+                break
+                
+            case .failure(let error):
+                print(error)
+                break
+            }
+        }
```
This piece of code can be moved after `courseOutlineStream` code block.

saeedbashir commented on Source/en.lproj/Localizable.strings:
```diff
@@ -276,6 +276,34 @@
 "COURSEDATES.COURSE_END_DATE"="Course End Date";
 /* Course Date Event */
 "COURSEDATES.EVENT"="Event";
+/* Course Reset Date Title */
+"COURSEDATES.RESET_DATE.TITLE"="Course Dates";
+/* Course Reset Date Success Message */
+"COURSEDATES.RESET_DATE.SUCCESS_MESSAGE"="Your dates have been successfully shifted.";
+/* Course Reset Date Error Message */
+"COURSEDATES.RESET_DATE.ERROR_MESSAGE"="Your dates could not be shifted. Please try again.";
+/* Course Reset Date Tab Info Banner Header */
+"COURSEDATES.RESET_DATE.TAB_INFO_BANNER.HEADER"="We've built a suggested schedule to help you stay on track. ";
+/* Course Reset Date Tab Info Banner Body */
+"COURSEDATES.RESET_DATE.TAB_INFO_BANNER.BODY"="But don't worry—it's flexible so you can learn at your own pace. If you happen to fall behind on our suggested dates, you'll be able to adjust them to keep yourself on track.";
+/* Course Reset Date Upgrade To Complete Graded Banner Header */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_COMPLETE_GRADED_BANNER.HEADER"="You are auditing this course, ";
+/* Course Reset Date Upgrade To Complete Graded Banner Body */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_COMPLETE_GRADED_BANNER.BODY"="which means that you are unable to participate in graded assignments. To complete graded assignments as part of this course, you can upgrade today.";
+/* Course Reset Date Upgrade To Complete Graded Banner Button */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_COMPLETE_GRADED_BANNER.BUTTON"="Upgrade now";
+/* Course Reset Date Upgrade Reset Banner Header */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_RESET_BANNER.HEADER"="You are auditing this course, ";
+/* Course Reset Date Upgrade Reset Banner Body */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_RESET_BANNER.BODY"="which means that you are unable to participate in graded assignments. It looks like you missed some important deadlines based on our suggested schedule. To complete graded assignments as part of this course and shift the past due assignments into the future, you can upgrade today.";
+/* Course Reset Date Upgrade Reset Banner Button */
+"COURSEDATES.RESET_DATE.UPGRADE_TO_RESET_BANNER.BUTTON"="Upgrade to shift due dates";
+/* Course Reset Date Banner Header */
+"COURSEDATES.RESET_DATE.RESET_DATE_BANNER.HEADER"="It looks like you missed some important deadlines based on our suggested schedule. ";
+/* Course Reset Date Banner Body */
+"COURSEDATES.RESET_DATE.RESET_DATE_BANNER.BODY"="To keep yourself on track, you can update this schedule and shift the past due assignments into the future. Don’t worry—you won’t lose any of the progress you’ve made when you shift your due dates.";
+/* Course Reset Date Banner Button */
+"COURSEDATES.RESET_DATE.RESET_DATE_BANNER.BUTTON"="Shift due dates";
```
Can you please update from where you have taken these strings?

mumer92 replied:
These Strings are picked up from here

https://github.com/edx/frontend-app-learning/blob/a6edc9132f648890eb5693e644998174894fbe41/src/course-home/dates-banner/messages.js#L3-L64

saeedbashir reviewed (changes requested):
* On successful dates shifting, the banner should refresh itself with the latest information.
* Color is not matching with the design.
* Header text is not bold as per the design it should be bold.
* There is a glitch in some course when the dates banner is there.
![ezgif-7-60bb04aa1a47](https://user-images.githubusercontent.com/5606473/93187066-cc994180-f758-11ea-9228-9921586b3729.gif)

mumer92 commented on Source/CourseDates.swift:
```diff
@@ -158,6 +212,36 @@ struct DatesBannerInfo {
+        return nil
+    }
+    
+    var showDatesTabBannerInfo: Bool {
+        return !missedDeadline
+    }
+    
+    var upgradeToCompleteGraded: Bool {
+        return contentTypeGatingEnabled && !missedDeadline
+    }
+    
+    var upgradeToReset: Bool {
+        return !upgradeToCompleteGraded && missedDeadline && missedGatedContent
+    }
+    
+    var resetDates: Bool {
+        return !upgradeToCompleteGraded && missedDeadline && !missedGatedContent
+    }
```
according to this [comment](https://openedx.atlassian.net/browse/LEARNER-7724?focusedCommentId=490264) and this [comment](https://openedx.atlassian.net/browse/LEARNER-7724?focusedCommentId=479326), we can ignore `courseDateBlocks`, `model ==='dates'` and `isSelfPaced` can be skipped

mumer92 commented on Source/CourseDatesAPI.swift:
```diff
@@ -19,16 +19,59 @@ public class CourseDatesAPI: NSObject {
     
     class func courseDatesRequest(courseID: String)-> NetworkRequest<CourseDateModel> {
-        let datesPath = path(courseID: courseID)
+        let datesPath = courseDatesPath(courseID: courseID)
```
There are other paths added as well, so gave this function appropriate name.

mumer92 commented:
@saeedbashir 
I have addressed the feedback, please review.

>Header text is not bold as per the design it should be bold.

As for bold text, seems like fonts are broken app wide, we need to investigate this in separate ticket.

saeedbashir commented on Source/CourseOutlineViewController.swift:
```diff
@@ -291,6 +318,24 @@ public class CourseOutlineViewController :
+    
+    private func reloadAfterCourseDateReset() {
+        courseQuerier.needsRefresh = true
+        addListeners()
+        loadStreams()
```
Do we need to add listeners again? to reload the data?

saeedbashir reviewed (changes requested):
* Can you please reduce the top and bottom empty space. Maybe to half of what it is now.
![Simulator Screen Shot - iPhone 11 Pro Max - 2020-09-16 at 10 08 22](https://user-images.githubusercontent.com/5606473/93294755-d2e4f780-f804-11ea-9232-9800c7efaa56.png)
* Dates banner is overlapping with the course card image. 
![Uploading ezgif-7-1bd40ce65a96.gif…]()
* We need to inform learners how ones can shift the dates. If this doesn't fall in the scope of this ticket then please create a followup ticket to address it.
<img width="364" alt="Screen Shot 2020-09-16 at 10 24 42 AM" src="https://user-images.githubusercontent.com/5606473/93295790-4c7de500-f807-11ea-892a-801f5f12adee.png">

mumer92 commented:
@saeedbashir 
I have moved some to code to functions inside `CourseOutlineViewController`, please see them is accordance to original file
https://github.com/edx/edx-app-ios/blob/master/Source/CourseOutlineViewController.swift

saeedbashir commented on Source/CourseResetDateBannerView.swift:
```diff
@@ -124,12 +152,19 @@ class CourseResetDateBannerView: UIView {
     private func setAccessibilityIdentifiers() {
-        messageLabel.accessibilityIdentifier = "CourseResetDateView:banner-text-label"
-        bannerButton.accessibilityIdentifier = "CourseResetDateView:reset-date-button"
+        messageLabel.accessibilityIdentifier = "CourseResetDateBannerView:banner-text-label"
+        bannerButton.accessibilityIdentifier = "CourseResetDateBannerView:reset-date-button"
```
It would be nice if you can assign accessibility identifiers to all the UIElements. 
Message label identifiers can be updated to `CourseResetDateBannerView:message-label` and likewise banner button and other elements.

saeedbashir reviewed (commented):
Please reload the banner status with `PullToRefresh`.

mumer92 commented:
> * We need to inform learners how ones can shift the dates. If this doesn't fall in the scope of this ticket then please create a followup ticket to address it.

Maybe we can add this to Whats New to inform user that they can reset dates.

saeedbashir commented:
> > * We need to inform learners how ones can shift the dates. If this doesn't fall in the scope of this ticket then please create a followup ticket to address it.
> 
> Maybe we can add this to Whats New to inform users that they can reset dates.

It is better to show a popup message on the click of a button, how the learner can shift the dates.

mumer92 commented:
> > > * We need to inform learners how ones can shift the dates. If this doesn't fall in the scope of this ticket then please create a followup ticket to address it.
> > 
> > 
> > Maybe we can add this to Whats New to inform users that they can reset dates.
> 
> It is better to show a popup message on the click of a button, how the learner can shift the dates.

@saeedbashir button is only visible when user can shift dates, in following case

https://github.com/edx/edx-app-ios/blob/6af2b9e56e1e090f6bf58ad8129f1f77c83831b3/Source/CourseDates.swift#L246-L250

and I think the banner text explanatory. 

https://github.com/edx/edx-app-ios/blob/6af2b9e56e1e090f6bf58ad8129f1f77c83831b3/Source/en.lproj/Localizable.strings#L301-L306

saeedbashir commented on Source/CourseResetDateBannerView.swift:
```diff
@@ -0,0 +1,213 @@
+private let cornerRadius: CGFloat = 5
+
+protocol CourseResetDateViewDelegate {
+    func courseResetButtonAction()
```
Isn't this reset the dates instead of course, please rename it accordingly.

saeedbashir commented on Source/CourseResetDateBannerView.swift:
```diff
@@ -0,0 +1,213 @@
+        guard let bannerInfo = bannerInfo, let status = bannerInfo.status else { return 0 }
+        
+        switch status {
+        case .upgradeToCompleteGradedBanner:
+            return 10
```
Any reason to return `10` only in case of `upgradeToCompleteGradedBanner`?

mumer92 replied:
Font is behaving weirdly, it is applying some padding to text, sometimes it applies and sometimes not, so in case of `upgradeToCompleteGradedBanner` it is fitting true to size, so added an offset to give a little space to label.

saeedbashir reviewed (commented):
If `hasEnded ` of `CourseDateDeadlineInfoModel` returns `true` on `PullToRefresh` then the dates header is not hiding itself.

saeedbashir commented on Source/CourseBannerAPI.swift:
```diff
@@ -0,0 +1,54 @@
+        return Failure()
+    }
+    
+    private class func resetDatePath(courseID: String) -> String {
```
This is being used in `courseBannerRequest` so according to the usage and functionality, it should be renamed to `path` or `datesPath` or to more relevant name.

saeedbashir commented on Source/CourseBannerAPI.swift:
```diff
@@ -0,0 +1,54 @@
+        return Failure()
+    }
+    
+    private class var datesShiftPath: String {
```
As it is used in the `courseDatesResetRequest` so it should be renamed accordingly.

saeedbashir commented on Source/CourseDatesAPI.swift:
```diff
@@ -19,15 +19,14 @@ public class CourseDatesAPI: NSObject {
     }
     
-    private class func path(courseID: String)-> String {
+    private class func courseDatesPath(courseID: String)-> String {
```
Now the other APIs moved to a separate file so I believe this change isn't required.

saeedbashir commented on Source/CourseOutlineTableSource.swift:
```diff
@@ -321,40 +332,76 @@ class CourseOutlineTableController : UITableViewController, CourseVideoTableView
+        updateResetBannerView(show: false)
+    }
+    
+    private func updateResetBannerView(show: Bool) {
```
It can be renamed to `updateCourseDateBannerView`.

saeedbashir reviewed (approved):
* Please create a followup ticket to add the screenshot tests for the course date banner. The PR LGTM 👍

salman2013 commented on Source/CourseOutlineViewController.swift:
```diff
@@ -154,29 +155,65 @@ public class CourseOutlineViewController :
+    }
+    
     private func reload() {
         self.blockIDStream.backWithStream(OEXStream(value : self.blockID))
```
self can be omitted.

salman2013 commented on Source/CourseOutlineViewController.swift:
```diff
@@ -213,35 +250,44 @@ public class CourseOutlineViewController :
         )
         
         self.blockIDStream.backWithStream(OEXStream(value: rootID))
```
self can be omitted

salman2013 commented on Source/CourseOutlineViewController.swift:
```diff
@@ -213,35 +250,44 @@ public class CourseOutlineViewController :
+    
+    private func loadRowsStream() {
+        rowsLoader.listen(self, success : { [weak self] groups in
+                if let owner = self {
```
Why we are wrapping self into variable here?

mumer92 replied:
on following line

https://github.com/edx/edx-app-ios/blob/dc4b7b52bb8251709524808de4371ec111631097/Source/CourseOutlineViewController.swift#L269

`owner.emptyState()` is used, so if it is changes to `self?.emptyState()`, its return type becomes `LoadState?` instead of `LoadState`

<img width="957" alt="Screenshot 2020-09-23 at 3 15 21 PM" src="https://user-images.githubusercontent.com/960241/93999725-9b9bbb00-fdaf-11ea-9d44-2a2ea5896ce0.png">

salman2013 commented on Source/CourseOutlineViewController.swift:
```diff
@@ -291,10 +337,31 @@ public class CourseOutlineViewController :
+                UIAlertController().showAlert(withTitle: Strings.Coursedates.ResetDate.title, message: Strings.Coursedates.ResetDate.errorMessage, onViewController: controller)
+            } else {
+                UIAlertController().showAlert(withTitle: Strings.Coursedates.ResetDate.title, message: Strings.Coursedates.ResetDate.successMessage, onViewController: controller)
+                self?.reloadAfterCourseDateReset()
```
I think we should also reload the view of the dates tab because when we shift dates the course screen update banner but the dates tab still shows the dates in past due.

mumer92 replied:
I think we need to handle that part in the other ticket which is relevant to adding banner view on DatesTab.

salman2013 replied:
@saeedbashir what do you think about this?

mumer92 replied:
I have made this change in this [commit](https://github.com/edx/edx-app-ios/pull/1407/commits/6d510d74399842d4f90cf8c521be0482c28d88f4)

mumer92 commented on Source/CourseDates.swift:
```diff
@@ -158,6 +212,41 @@ struct DatesBannerInfo {
+    
+    // Case 4
+    private var resetDates: Bool {
+        return !upgradeToCompleteGraded && missedDeadline && !missedGatedContent
```
This logic mirrors the logic that is implemented on MFE, here is the relevant piece

https://github.com/edx/frontend-app-learning/blob/master/src/course-home/dates-banner/DatesBannerContainer.jsx#L40

` const resetDates = !upgradeToCompleteGraded && missedDeadlines && !missedGatedContent;`

mumer92 replied:
We also need to check `missedDeadline` here as too.

salman2013 replied:
My point is if we are using the property "upgradeToCompleteGraded" which is actually returning `contentTypeGatingEnabled && !missedDeadline` so there is no need to check `missedDeadlines ` again

salman2013 replied:
try to test it by doing it.

mumer92 replied:
Please check following:

```
    let contentTypeGatingEnabled = false
    let missedDeadline = false
    let missedGatedContent = false
    
    // returns false
    private var upgradeToCompleteGraded: Bool {
        return contentTypeGatingEnabled && !missedDeadline
    }

    private var resetDates: Bool {
        !upgradeToCompleteGraded && missedDeadline && !missedGatedContent    // returns false
        !upgradeToCompleteGraded && !missedGatedContent    // returns true
    }

```

salman2013 replied:
Then i think we should use it like 
`(!contentTypeGatingEnabled && missedDeadline && !missedDeadline`)
 because 

`upgradeToCompleteGraded ` mean `contentTypeGatingEnabled && !missedDeadline`
and we are checking not `upgradeToCompleteGraded` which means the `!contentTypeGatingEnabled && missedDeadline`` 

What I am understanding if we are saying  **! upgradeToCompleteGraded** is false then we already mean missedDeadlines is true so no logical reason to check it as condition again.

mumer92 replied:
>Then i think we should use it like
(!contentTypeGatingEnabled && missedDeadline && !missedDeadline)

I believe you meant `!contentTypeGatingEnabled && missedDeadline && !missedGatedContent`

Lets assume the following:
```
let contentTypeGatingEnabled = true
let missedDeadline = true
let missedGatedContent = false

    // returns false
private var upgradeToCompleteGraded: Bool {
    return contentTypeGatingEnabled && !missedDeadline
}

private var resetDates: Bool {
    !upgradeToCompleteGraded && missedDeadline && !missedGatedContent    // returns true
    !contentTypeGatingEnabled && missedDeadline && !missedGatedContent   // returns false
}
``` 

I believe `upgradeToCompleteGraded` is separate check that ensures that a user is allowed to upgrade to complete gates content and the actual case for which banner must be shown, have different condition which accumulates `upgradeToCompleteGraded`, `missedDeadline` and `missedGatedContent` and based on their values, decides if banner for shifting dates must be shown to user or not.

mumer92 commented on Source/CourseDates.swift:
```diff
@@ -158,6 +212,41 @@ struct DatesBannerInfo {
+    
+    // Case 3
+    private var upgradeToReset: Bool {
+        return !upgradeToCompleteGraded && missedDeadline && missedGatedContent
```
This logic mirrors the logic that is implemented on MFE, here is the relevant piece

https://github.com/edx/frontend-app-learning/blob/master/src/course-home/dates-banner/DatesBannerContainer.jsx#L39

`const upgradeToReset = !upgradeToCompleteGraded && missedDeadlines && missedGatedContent;`

mumer92 replied:
We also need to check `missedDeadline` here as too.

mumer92 replied:
Lets say 
`contentTypeGatingEnabled = false` and `missedDeadline = false`
so `upgradeToReset` should return `contentTypeGatingEnabled && !missedDeadline` which will be 
`false && !false` which will be equal to `false`

now in `upgradeToReset : Bool`  if I eliminate, `!missedDeadline` so expression becomes 
`!upgradeToCompleteGraded && missedGatedContent`

Suppose, 
`upgradeToCompleteGraded  = false` and `missedGatedContent = true`
then 
`!upgradeToCompleteGraded && missedGatedContent` will return `true`

Lets introduce `missedDeadline = false`, which in above lines, was also `false`, then the expression:
`!upgradeToCompleteGraded && missedDeadline && missedGatedContent` will return `false`

To put it in perspective, 

```
    let missedDeadline = false
    let missedGatedContent = true
    let contentTypeGatingEnabled = false

   // returns false
   private var upgradeToCompleteGraded: Bool {
        return contentTypeGatingEnabled && !missedDeadline
   }

    private var upgradeToReset: Bool {
        !upgradeToCompleteGraded && missedDeadline && missedGatedContent    // returns false
        !upgradeToCompleteGraded && missedGatedContent    // returns true
    }

```

mumer92 commented on Source/CourseDateBannerView.swift:
```diff
@@ -0,0 +1,203 @@
+
+import UIKit
+
+private let cornerRadius: CGFloat = 5
```
I think it should be fine, as `cornerRadius` already has a type of  `CGFloat` so it treats `5` itself by typecasting it.

salman2013 commented:
I have one question, After shifting the dates all sections are showing future dates like of month October but the course ending date is still the same. is that the correct scenario?

![Simulator Screen Shot - iPhone 11 Pro - 2020-09-23 at 15 35 42](https://user-images.githubusercontent.com/4454296/94001810-7f4d4d80-fdb2-11ea-9ba1-507cd0c90208.png)

mumer92 commented on Source/CourseDateBannerModel.swift:
```diff
@@ -0,0 +1,28 @@
+
+import Foundation
+
+let NOTIFICATION_SHIFT_COURSE_DATES_SUCCESS_FROM_COURSE_DASHBOARD = "ShifCourseDatesSuccessFromCourseDashboard"
```
done

salman2013 reviewed (approved):
LGTM
